#!/bin/bash
#PBS -P pq84
#PBS -q normal 
#PBS -N Variant_calling_bcftools
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=128GB,ncpus=16
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"
echo "Chromosomes 9 to 12"

echo "---------------------------------------"
echo "Define paths and load modules"
module load bcftools/1.12

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
INDEXTDIR="/g/data/pq84/malaria/Pf_Malaysia/data/ref_genomes/PlasmoDB-59_Pfalciparum3D7_Genome.fasta"
OUTDIR="/g/data/pq84/malaria/Pf_Malaysia/outputs/04_Variant_calling/bcftools"
INDIR="/g/data/pq84/malaria/Pf_Malaysia/outputs/03_Bam-pre/final_bam"

echo "---------------------------------------------------------------------------------------------------------------------"
echo "create input bam file list"
echo "---------------------------------------------------------------------------------------------------------------------"
cd $INDIR

echo "---------------------------------------------------------------------------------------------------------------------"
echo "bcftools"
echo "---------------------------------------------------------------------------------------------------------------------"

bcftools mpileup --threads 16 -f $INDEXTDIR -b $INDIR/input_bam_files.list \
   -r Pf3D7_09_v3,Pf3D7_10_v3,Pf3D7_11_v3,Pf3D7_12_v3 \
   | bcftools call --threads 16 -m -Oz -a FORMAT/GQ,FORMAT/GP,INFO/PV4 -v -o $OUTDIR/bcftools_genotyped_3.vcf.gz

bcftools index --threads 16 -t -o $OUTDIR/bcftools_genotyped_3.vcf.gz.tbi $OUTDIR/bcftools_genotyped_3.vcf.gz

echo "---------------------------------------"
echo "Finished "
