#!/bin/bash
#PBS -P pq84
#PBS -q normalbw
#PBS -N Variant_calling_gatk
#PBS -j oe
#PBS -m ae
#PBS -l walltime=48:00:00,mem=128GB,ncpus=16
#PBS -l storage=gdata/pq84+scratch/pq84
#PBS -M jacob.westaway@menzies.edu.au

echo "---------------------------------------"
echo "PBS: Job identifier is $PBS_JOBID"
echo "PBS: Job name is $PBS_JOBNAME"

echo "---------------------------------------"
echo "Define paths and load modules"
module load java/jdk-8.40 

echo "---------------------------------------"
echo "---------------------------------------"
echo 'Change to working directory and set env variables'
OUTDIR="/g/data/pq84/malaria/Pf_Malaysia/outputs/04_Variant_calling/gatk"
INDIR="/g/data/pq84/malaria/Pf_Malaysia/outputs/04_Variant_calling/haplotype_call"
PICARD="/g/data/pq84/bin/picard/build/libs/picard.jar"
GATK="/g/data/pq84/bin/GenomeAnalysisTK-3.8-1-0/GenomeAnalysisTK.jar"
INDEXTDIR="/g/data/pq84/malaria/Pf_Malaysia/data/ref_genomes/PlasmoDB-59_Pfalciparum3D7_Genome.fasta"
INTERVALS="/g/data/pq84/malaria/Pf_Malaysia/data/ref_genomes/PlasmoDB-59_Pfalciparum3D7_Genome.fasta.bed"

cd $INDIR

echo "---------------------------------------------------------------------------------------------------------------------"
echo "GATK"
echo "---------------------------------------------------------------------------------------------------------------------"

echo "---------------------------------------------------------------------------------------------------------------------"
echo "Combine GVCFs"

java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
    -T CombineGVCFs \
    -R $INDEXTDIR \
    -V ERR9907878.dupmarked.realigned.recal.g.vcf.gz -V ERR9907879.dupmarked.realigned.recal.g.vcf.gz -V ERR9907880.dupmarked.realigned.recal.g.vcf.gz -V ERR9907881.dupmarked.realigned.recal.g.vcf.gz -V ERR9907882.dupmarked.realigned.recal.g.vcf.gz -V ERR9907883.dupmarked.realigned.recal.g.vcf.gz -V ERR9907884.dupmarked.realigned.recal.g.vcf.gz -V ERR9907885.dupmarked.realigned.recal.g.vcf.gz -V ERR9907886.dupmarked.realigned.recal.g.vcf.gz -V ERR9907887.dupmarked.realigned.recal.g.vcf.gz -V ERR9907888.dupmarked.realigned.recal.g.vcf.gz -V ERR9907889.dupmarked.realigned.recal.g.vcf.gz -V ERR9907890.dupmarked.realigned.recal.g.vcf.gz -V ERR9907891.dupmarked.realigned.recal.g.vcf.gz -V ERR9907892.dupmarked.realigned.recal.g.vcf.gz -V ERR9907893.dupmarked.realigned.recal.g.vcf.gz -V ERR9907894.dupmarked.realigned.recal.g.vcf.gz -V ERR9907895.dupmarked.realigned.recal.g.vcf.gz -V ERR9907896.dupmarked.realigned.recal.g.vcf.gz -V ERR9907897.dupmarked.realigned.recal.g.vcf.gz -V ERR9907898.dupmarked.realigned.recal.g.vcf.gz -V ERR9907899.dupmarked.realigned.recal.g.vcf.gz -V ERR9907900.dupmarked.realigned.recal.g.vcf.gz -V ERR9907901.dupmarked.realigned.recal.g.vcf.gz -V ERR9907902.dupmarked.realigned.recal.g.vcf.gz -V ERR9907903.dupmarked.realigned.recal.g.vcf.gz -V ERR9907904.dupmarked.realigned.recal.g.vcf.gz -V ERR9907905.dupmarked.realigned.recal.g.vcf.gz -V ERR9907906.dupmarked.realigned.recal.g.vcf.gz -V ERR9907907.dupmarked.realigned.recal.g.vcf.gz -V ERR9907908.dupmarked.realigned.recal.g.vcf.gz -V ERR9907909.dupmarked.realigned.recal.g.vcf.gz -V ERR9907910.dupmarked.realigned.recal.g.vcf.gz -V ERR9907911.dupmarked.realigned.recal.g.vcf.gz -V ERR9907912.dupmarked.realigned.recal.g.vcf.gz -V ERR9907913.dupmarked.realigned.recal.g.vcf.gz -V ERR9907914.dupmarked.realigned.recal.g.vcf.gz -V ERR9907915.dupmarked.realigned.recal.g.vcf.gz -V ERR9907916.dupmarked.realigned.recal.g.vcf.gz -V ERR9907917.dupmarked.realigned.recal.g.vcf.gz -V ERR9907918.dupmarked.realigned.recal.g.vcf.gz -V ERR9907919.dupmarked.realigned.recal.g.vcf.gz -V ERR9907920.dupmarked.realigned.recal.g.vcf.gz -V ERR9907921.dupmarked.realigned.recal.g.vcf.gz -V ERR9907922.dupmarked.realigned.recal.g.vcf.gz -V ERR9907923.dupmarked.realigned.recal.g.vcf.gz -V ERR9907924.dupmarked.realigned.recal.g.vcf.gz -V ERR9907925.dupmarked.realigned.recal.g.vcf.gz -V ERR9937125.dupmarked.realigned.recal.g.vcf.gz -V ERR9937126.dupmarked.realigned.recal.g.vcf.gz -V ERR9937127.dupmarked.realigned.recal.g.vcf.gz -V ERR9937128.dupmarked.realigned.recal.g.vcf.gz -V ERR9937129.dupmarked.realigned.recal.g.vcf.gz -V ERR9937130.dupmarked.realigned.recal.g.vcf.gz -V ERR9937131.dupmarked.realigned.recal.g.vcf.gz -V ERR9937132.dupmarked.realigned.recal.g.vcf.gz -V ERR9937133.dupmarked.realigned.recal.g.vcf.gz -V ERR9937134.dupmarked.realigned.recal.g.vcf.gz -V ERR9937135.dupmarked.realigned.recal.g.vcf.gz -V ERR9937136.dupmarked.realigned.recal.g.vcf.gz -V ERR9937137.dupmarked.realigned.recal.g.vcf.gz -V ERR9937138.dupmarked.realigned.recal.g.vcf.gz -V ERR9937139.dupmarked.realigned.recal.g.vcf.gz -V ERR9937140.dupmarked.realigned.recal.g.vcf.gz -V ERR9937141.dupmarked.realigned.recal.g.vcf.gz -V ERR9937142.dupmarked.realigned.recal.g.vcf.gz -V ERR9937143.dupmarked.realigned.recal.g.vcf.gz -V ERR9937144.dupmarked.realigned.recal.g.vcf.gz -V ERR9937145.dupmarked.realigned.recal.g.vcf.gz -V ERR9937146.dupmarked.realigned.recal.g.vcf.gz -V ERR9937147.dupmarked.realigned.recal.g.vcf.gz -V ERR9937148.dupmarked.realigned.recal.g.vcf.gz \
    -o $OUTDIR/GATK_combined.g.vcf.gz

echo "---------------------------------------------------------------------------------------------------------------------"
echo "Genotype GVCFs"

cd $OUTDIR

java -Djava.iodir=$PBS_JOBFS -Xms3200m -Xmx3600m -jar $GATK \
    -T GenotypeGVCFs \
    -nt 16 \
    -R $INDEXTDIR \
    -L $INTERVALS \
    -V GATK_combined.g.vcf.gz \
    -o GATK_genotyped.vcf.gz

echo "---------------------------------------"
echo "Finished "
